sw/device/tests/dif/dif_uart_test.cc

```diff
@@ -0,0 +1,657 @@
+                                     &tx_watermark_state));
+
+  EXPECT_EQ(tx_watermark_state, kDifUartEnable);
+
+  // Get the last IRQ state.
+  EXPECT_READ32(UART_INTR_STATE_REG_OFFSET,
+                {{UART_INTR_STATE_RX_PARITY_ERR, false}});
+
+  dif_uart_enable_t last_state = kDifUartEnable;
+  EXPECT_TRUE(
+      dif_uart_irq_state_get(&dif_uart_, kDifUartInterruptLast, &last_state));
+
+  EXPECT_EQ(last_state, kDifUartDisable);
+}
+
+class IrqStateClearTest : public UartTest {};
+
+TEST_F(IrqStateClearTest, UartNull) {
+  EXPECT_FALSE(dif_uart_irq_state_clear(nullptr, kDifUartInterruptTxWatermark));
+}
+
+TEST_F(IrqStateClearTest, Success) {
+  // Clear the first IRQ state.
+  EXPECT_MASK32(UART_INTR_STATE_REG_OFFSET,
+                {{UART_INTR_STATE_TX_WATERMARK, 0x1, true}});
+
+  EXPECT_TRUE(
+      dif_uart_irq_state_clear(&dif_uart_, kDifUartInterruptTxWatermark));
+
+  // Clear the last IRQ state.
+  EXPECT_MASK32(UART_INTR_STATE_REG_OFFSET,
+                {{UART_INTR_STATE_RX_PARITY_ERR, 0x1, true}});
+
+  EXPECT_TRUE(dif_uart_irq_state_clear(&dif_uart_, kDifUartInterruptLast));
+}
+
+class IrqsDisableTest : public UartTest {};
+
+TEST_F(IrqsDisableTest, UartNull) {
+  uint32_t state;
+  EXPECT_FALSE(dif_uart_irqs_disable(nullptr, &state));
+}
+
+TEST_F(IrqsDisableTest, Success) {
+  EXPECT_WRITE32(UART_INTR_ENABLE_REG_OFFSET, 0);
+
+  EXPECT_TRUE(dif_uart_irqs_disable(&dif_uart_, nullptr));
+}
+
+TEST_F(IrqsDisableTest, AllDisabled) {
+  // IRQs disabled
+  EXPECT_READ32(UART_INTR_ENABLE_REG_OFFSET, 0);
+  EXPECT_WRITE32(UART_INTR_ENABLE_REG_OFFSET, 0);
+
+  uint32_t state = std::numeric_limits<uint32_t>::max();
+  EXPECT_TRUE(dif_uart_irqs_disable(&dif_uart_, &state));
+
+  EXPECT_EQ(state, 0);
+}
+
+TEST_F(IrqsDisableTest, AllEnabled) {
+  // IRQs enabled
+  EXPECT_READ32(UART_INTR_ENABLE_REG_OFFSET,
+                std::numeric_limits<uint32_t>::max());
+  EXPECT_WRITE32(UART_INTR_ENABLE_REG_OFFSET, 0);
+
+  uint32_t state = 0;
+  EXPECT_TRUE(dif_uart_irqs_disable(&dif_uart_, &state));
+
+  EXPECT_EQ(state, std::numeric_limits<uint32_t>::max());
+}
+
+class IrqsRestoreTest : public UartTest {};
+
+TEST_F(IrqsRestoreTest, UartNull) {
+  EXPECT_FALSE(
+      dif_uart_irqs_restore(nullptr, std::numeric_limits<uint32_t>::max()));
+}
+
+TEST_F(IrqsRestoreTest, AllEnabled) {
+  uint32_t state = std::numeric_limits<uint32_t>::max();
+  EXPECT_WRITE32(UART_INTR_ENABLE_REG_OFFSET, state);
+
+  EXPECT_TRUE(dif_uart_irqs_restore(&dif_uart_, state));
+}
+
+TEST_F(IrqsRestoreTest, NoneEnabled) {
+  uint32_t state = 0;
+  EXPECT_WRITE32(UART_INTR_ENABLE_REG_OFFSET, state);
+
+  EXPECT_TRUE(dif_uart_irqs_restore(&dif_uart_, state));
+}
+
+class IrqEnableTest : public UartTest {};
+
+TEST_F(IrqEnableTest, UartNull) {
+  EXPECT_FALSE(dif_uart_irq_enable(nullptr, kDifUartInterruptTxWatermark,
+                                   kDifUartEnable));
+}
+
+TEST_F(IrqEnableTest, Success) {
```

Comment:
There are 9 options but this test exercises only 2 of them. IMHO, remaining 7 also deserve to be tested :)

Reply:
Initially I was testing all of them, although when I came to my PLIC unittests it proved to be problematic and there is no good way to test every bit of multireg type of IE, IP, LE registers without making the tests more complicated.

There are trade-offs with every approach. The reason why I went with simply testing the last and the first variant because the API follows the same logic for all of the variants. Yes, there is more chance of missing a bug, but I think it is a reasonable compromise for now, and we can come back to it later if need to.

Reply:
That is exactly the reason why I would love to see things like `UART_INTR_STATE_RX_WATERMARK` as an enum, so that it is possible to simply loop through it. And it should be easily doable for peripherals like UART (the IRQs are internal information to the peripheral), it is much trickier with peripherals like PLIC, as a lot of information comes from `top_earlgrey`.